modules/effects/arrow-effects/src/main/kotlin/arrow/effects/internal/Promise.kt

```diff
@@ -0,0 +1,104 @@
+package arrow.effects.internal
+
+import arrow.Kind
+import arrow.core.*
+import arrow.effects.Promise
+import java.util.concurrent.atomic.AtomicReference
+
+fun <A> Promise.Companion.unsafe(): Promise<ForId, A> = UnsafePromise()
+
+internal class UnsafePromise<A> : Promise<ForId, A> {
+
+  private val state: AtomicReference<State<A>> = AtomicReference(State.Pending(emptyList()))
+
+  override val get: Id<A>
+    get() {
+      tailrec fun loop(): Id<A> = when (val st = state.get()) {
+        is State.Pending<A> -> loop()
+        is State.Full -> Id(st.value)
+        is State.Error -> throw st.throwable
+      }
+
+      tailrec fun calculateNewState(): Unit {
+        val oldState = state.get()
+        val newState = when (oldState) {
+          is State.Pending<A> -> State.Pending(oldState.joiners)
+          is State.Full -> oldState
+          is State.Error -> oldState
+        }
+        return if (state.compareAndSet(oldState, newState)) Unit else calculateNewState()
+      }
+
+      calculateNewState()
+      return loop()
+    }
+
+  override val tryGet: Kind<ForId, Option<A>>
+    get() = when (val oldState = state.get()) {
+      is State.Pending<A> -> Id(None)
+      is State.Full -> Id(Some(oldState.value))
+      is State.Error -> Id(None)
+    }
+
+  override fun tryComplete(a: A): Id<Boolean> {
+    tailrec fun calculateNewState(): Unit {
```

Comment:
This creates a new lambda on every execution. Better move it outside and mark it private.